File: src_fortran/sub/inputcnv.f
c
c
c
c
c
c
      subroutine INPUTCNV(iunit,nobs,
     &                  sta,iyr,imo,iday,ihr,imin,sec,
     &                  rmk1,rmk2,cphase,iwt,amx,prx,
     &                  xlat,xlon,xmagni,depth,origtime,
     &                  i1,i2,i3,eventtype)
c
c     on Output:
c     ---------
c     EQS:   i1=ifx(i)   
c     SHOTS: i1=icc
c
      implicit none
c
      integer maxobsperevent
      parameter (maxobsperevent=180)
      integer nobs
      character*4 sta(maxobsperevent)
      integer*4 itime
      real sec(maxobsperevent)
      character*1 rmk1(maxobsperevent),rmk2(maxobsperevent),
     &            cphase(maxobsperevent)
      integer iunit,iwt(maxobsperevent)
      real amx(maxobsperevent), prx(maxobsperevent),
     &     xlat,xlon
      real xmagni, depth, origtime
      real ttime(maxobsperevent)
      character*1 eventtype
      integer i1,i2,i3
      character*1 cns,cew
c
      character*80 cline
      integer iyr,imo,iday,ihr,imin
      integer j,j1,j2
c
  1   eventtype='L'
      j2=-1
      nobs=0
c
   2  read(iunit,'(a)',end=99) cline
      if(cline.eq.' ') goto 2
cek search for end of *.cnv file, marked by 9999
      if(cline.eq.'9999') goto 999
c
cek   next line adjusted by EK 3.12.90
      read(cline,'(3i2,1x,2i2,1x,f5.2,1x,f7.4,a1,1x,f8.4,a1,1x,f7.2,
     1      2x,f5.2)',
     &           err=9999,end=999)
     &           iyr,imo,iday,ihr,imin,origtime,
     &           xlat,cns,xlon,cew,depth,xmagni ! i1 =ifx(i)
                !=======================!   trial hypocenter !!
cek  i1-switch is dummy
      i1=0
c
cek     EQS:   i1=ifx(i)     ! no longer in use !!!
c     SHOTS: i1=icc
c
      if(cns.eq.'S') xlat=-xlat
      if(cew.eq.'E') xlon=-xlon
c
      call TIMECLEAR(iyr,imo,iday,ihr,imin,origtime,itime)
c
      j2=0
   10 j1=j2+1
      j2=j1+5
      read(iunit,'(6(a4,a1,i1,f6.2))',end=999)
     &            (sta(j),cphase(j),iwt(j),ttime(j),j=j1,j2)
      do j=j1,j2
         if(sta(j).eq.' ')then
            if(j.eq.1.and.j1.eq.1) goto 1  ! blank-line read !
            j2=j-1   ! event is completely read-in
            goto 99
         endif
      enddo
      goto 10         ! read next input-line
c
   99 continue
      nobs=j2
      do j=1,nobs
         sec(j)=origtime+ttime(j)
      enddo
c
      RETURN
c
  999 continue
      nobs=-1
      RETURN
c
 9999 continue
      write(6,*)'INPUTCNV>>> read-error! input-line is:'
      write(6,'(a)') cline
      stop'subr. INPUTCNV >>> error!'
c
      end ! of subr. inputcnv

File: src/sub/inputcnv.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

extern void timeclear(int *iyr, int *imo, int *iday, int *ihr, int *imin,
                      double *sec, int *itime);

static int is_blank_line(const char *s) {
    if (s == NULL) return 1;
    for (const unsigned char *p = (const unsigned char *)s; *p; ++p) {
        if (!isspace(*p)) return 0;
    }
    return 1;
}

static void slice_copy(char *dst, size_t dst_size, const char *src, int start, int len) {
    if (dst == NULL || dst_size == 0) return;
    int i = 0;
    for (; i < len && (size_t)i + 1 < dst_size && src[start + i] != '\0'; ++i) {
        dst[i] = src[start + i];
    }
    dst[i] = '\0';
}

static void rstrip_inplace(char *s) {
    if (s == NULL) return;
    size_t n = strlen(s);
    while (n > 0 && isspace((unsigned char)s[n - 1])) {
        s[--n] = '\0';
    }
}

void inputcnv(FILE *fp, int *nobs,
              char sta[][5], int *iyr, int *imo, int *iday, int *ihr, int *imin, double sec[],
              char rmk1[], char rmk2[], char cphase[], int iwt[], double amx[], double prx[],
              double *xlat, double *xlon, double *xmagni, double *depth, double *origtime,
              int *i1, int *i2, int *i3, char *eventtype) {
    (void)rmk1;
    (void)rmk2;
    (void)amx;
    (void)prx;
    (void)i2;
    (void)i3;

    const int maxobsperevent = 180;
    char cline[256];
    double ttime[180] = {0.0};
    char cns = 'N';
    char cew = 'W';
    int itime;

    *eventtype = 'L';
    int j2 = -1;
    *nobs = 0;

read_event_header:
    while (fgets(cline, sizeof(cline), fp) != NULL) {
        if (is_blank_line(cline)) continue;
        if (strncmp(cline, "9999", 4) == 0) {
            *nobs = -1;
            return;
        }
        break;
    }

    if (feof(fp)) {
        *nobs = -1;
        return;
    }

    if (sscanf(cline,
               "%2d%2d%2d %2d%2d %5lf %7lf%c %8lf%c %7lf %5lf",
               iyr, imo, iday, ihr, imin, origtime,
               xlat, &cns, xlon, &cew, depth, xmagni) < 12) {
        fprintf(stderr, "INPUTCNV>>> read-error! input-line is:\n");
        fprintf(stderr, "%s", cline);
        exit(1);
    }

    *i1 = 0;
    if (cns == 'S') *xlat = -*xlat;
    if (cew == 'E') *xlon = -*xlon;

    timeclear(iyr, imo, iday, ihr, imin, origtime, &itime);

    while (1) {
        int j1 = j2 + 1;
        j2 = j1 + 5;

        if (fgets(cline, sizeof(cline), fp) == NULL) {
            break;
        }
        for (int k = 0; k < 6; ++k) {
            int j = j1 + k;
            if (j >= maxobsperevent) break;

            int off = 12 * k;
            char s4[8] = {0}, ph[4] = {0}, wt[8] = {0}, tt[16] = {0};
            slice_copy(s4, sizeof(s4), cline, off + 0, 4);
            slice_copy(ph, sizeof(ph), cline, off + 4, 1);
            slice_copy(wt, sizeof(wt), cline, off + 5, 1);
            slice_copy(tt, sizeof(tt), cline, off + 6, 6);

            rstrip_inplace(s4);

            if (s4[0] == '\0') {
                if (j == 0 && j1 == 0) {
                    goto read_event_header;
                }
                j2 = j - 1;
                goto done_obs;
            }

            strncpy(sta[j], s4, 4);
            sta[j][4] = '\0';
            cphase[j] = (ph[0] == '\0') ? ' ' : ph[0];
            iwt[j] = atoi(wt);
            ttime[j] = strtof(tt, NULL);
        }
    }

done_obs:
    *nobs = j2 + 1;
    if (*nobs < 0) {
        *nobs = -1;
        return;
    }

    for (int j = 0; j < *nobs; ++j) {
        sec[j] = *origtime + ttime[j];
    }
}
